libsolidity/ast/AST.cpp

```diff
@@ -643,20 +642,10 @@ set<VariableDeclaration::Location> VariableDeclaration::allowedDataLocations() c
 	}
 	else if (isLocalVariable())
 	{
-		solAssert(typeName(), "");
-		auto dataLocations = [](TypePointer _type, auto&& _recursion) -> set<Location> {
-			solAssert(_type, "Can only be called after reference resolution");
-			switch (_type->category())
-			{
-				case Type::Category::Array:
-					return _recursion(dynamic_cast<ArrayType const*>(_type)->baseType(), _recursion);
-				case Type::Category::Mapping:
-					return set<Location>{ Location::Storage };
-				default:
-					return set<Location>{ Location::Memory, Location::Storage, Location::CallData };
-			}
-		};
-		return dataLocations(typeName()->annotation().type, dataLocations);
+		if (typeName().annotation().type->containsNestedMapping())
```

Comment:
Can `typeName().annotation().type` be `nullptr` here?

Reply:
I see that the assert is moved to the constructor. `containsNestedMapping()` also has asserts if called before type resolution. So this should work.

Reply:
Do you think we can skip the test for `containsNestedMapping()` here? And simply return `set<Location>{ Location::Memory, Location::Storage, Location::CallData }` ? Then the type checker will be able to issue proper error messages. 

Since the mappings in structs were only disallowed since 0.7.0, I think this new error message would be very confusing.

Reply:
Good idea! I'll do it like that!